quinn/src/lib.rs

```diff
@@ -481,137 +414,91 @@ impl EndpointInner {
-        }
-        Ok(blocked)
-    }
-
-    fn drive_transmit(&mut self, now: Instant) -> Result<bool, io::Error> {
-        let mut sent = 0;
-        while let Some(x) = self.inner.poll_transmit(now) {
-            match self.socket.poll_send(&x.destination, x.ecn, &x.packet) {
-                Ok(Async::Ready(_)) => {}
-                Ok(Async::NotReady) => {
-                    self.outgoing = Some(x);
-                    break;
-                }
-                Err(ref e) if e.kind() == io::ErrorKind::PermissionDenied => {
-                    self.outgoing = Some(x);
-                    break;
-                }
-                Err(e) => {
-                    return Err(e);
-                }
-            }
-            sent += 1;
-            if sent >= IO_LOOP_BOUND {
+            if blocked {
                 return Ok(true);
             }
         }
-        Ok(false)
+        Ok(blocked)
     }
 
-    fn drive_timers(&mut self, now: Instant) -> Result<bool, io::Error> {
+    fn handle_events(&mut self) -> Result<bool, io::Error> {
+        use EndpointEvent::*;
         let mut keep_going = false;
         loop {
-            match self.timers.poll() {
-                Ok(Async::Ready(Some(Some((ch, timer))))) => {
-                    trace!(self.log, "timeout"; "timer" => ?timer);
-                    self.inner.timeout(now, ch, timer);
-                    if timer == quinn::Timer::Close {
-                        // Connection drained
-                        if let hash_map::Entry::Occupied(mut p) = self.pending.entry(ch) {
-                            if let Some(x) = p.get_mut().closing.take() {
-                                let _ = x.send(());
+            match self.events.poll() {
+                Ok(Async::Ready(Some((ch, event)))) => {
+                    match event {
+                        Connected => {
+                            self.connecting.remove(&ch);
+                        }
+                        Proto(e) => {
+                            if let quinn::EndpointEvent::Closed { .. } = &e {
+                                self.connections.remove(&ch);
                             }
-                            if p.get().dropped {
-                                p.remove();
-                            } else {
-                                p.get_mut().drained = true;
+                            if let Some(event) = self.inner.handle_event(ch, e) {
+                                if let hash_map::Entry::Occupied(mut entry) =
+                                    self.connections.entry(ch)
+                                {
+                                    entry.get_mut().unbounded_send(event).unwrap();
+                                }
                             }
                         }
+                        Transmit(t) => self.outgoing.push_back(t),
                     }
-                    // Timeout call may have queued sends
                     keep_going = true;
```

Comment:
A comment as to why this is necessary would be helpful.

Reply:
I don't think it is necessary? Removed it for now.